### ISSUE TITLE
Support repartitioning of topic data for stream-table joins more elegantly

### ISSUE THREAD

**Issue by rmoff** [confluentinc/ksql#2314]
KSQL currently requires a topic backing a Table to have message keys equal to the declared key of the table. If they are not, then the Table will not function as intended, with no error or warning. This causes confusion and frustration for the user: 

* https://github.com/confluentinc/ksql/issues/749
* https://github.com/confluentinc/ksql/issues/1405

This is a frequent occurance as keys on Kafka messages are not mandatory and may well not be set on a source topic. In addition, the target user for KSQL may not initially be familiar with the concept of keys in Kafka messages and so overlook this requirement entirely. 

Take an example in which it is desired to join a stream of events `wibble` to a table `FOO` of lookup details, on the common join column `BAR`. The source `foo` topic is not keyed. Done like this, the join would fail silently: 

    CREATE TABLE foo_table \
    WITH (KAFKA_TOPIC='foo', \
          VALUE_FORMAT='AVRO', \
          KEY='BAR');

    SELECT * FROM wibble w INNER JOIN foo_table f on w.BAR=f.BAR;

The join would fail silently because even though a `KEY` is declared in the `CREATE TABLE`, it does not match _the actual key of the Kafka topic_.  This is the case for both null-keyed messages, as well as messages with a key *but a key that does not match the value of the `BAR` column in the message value*.

 The current workaround is: 

1. Register the source topic as a stream: 

        CREATE STREAM foo_stream \
        WITH (KAFKA_TOPIC='foo', \
              VALUE_FORMAT='AVRO');

2. Repartition the topic on the desired key, writing to a new one: 

        CREATE STREAM FOO_REKEYED AS \
        SELECT * FROM foo \
        PARTITION BY BAR;

    Note that by repartitioning, all ordering guarentees are lost. 

3. Declare a table over the _rekeyed topic_: 

        CREATE TABLE foo_table2 \
        WITH (KAFKA_TOPIC='FOO_REKEYED', \
              VALUE_FORMAT='AVRO', \
              KEY='BAR');

Of these three steps, two are redundant. It should only be necessary to declare the `TABLE` and its key, and KSQL handle the rest. This is particularly true given that in the DDL `WITH` clause, `KEY` is specified which would leave most users to understand that KSQL will act on it, or at the very least throw an error if the message key does not match the declared one. 

Consideration would need to be given to the issue of order guarentees, so perhaps KSQL should have a DDL option to force the rekey, or to throw an error if the declared key doesn't match the message key.

Related discussion: 

* https://github.com/confluentinc/ksql/issues/804

**Comment by mjsax** [comment]
Seems to be related to https://github.com/confluentinc/ksql/issues/749  and I think we can close both tickets?
